### DISCUSSION TITLE
2.2 数据抽象

### DISCUSSION THREAD

**Discussion by giscus** [csfive/composing-programs-zh#107] [Announcements]
# 2.2 数据抽象

SICP Python 中文翻译

https://composingprograms.netlify.app/2/2

**Comment by EmbroiderSnow** [comment]
>>> from fractions import gcd

ImportError: cannot import name 'gcd' from 'fractions' 这个错误通常意味着你试图从 fractions 模块导入 gcd 函数，但是这个函数实际上并不在 fractions 模块中。在 Python 3.x 中，gcd（最大公约数）函数实际上是在 math 模块中，而不是 fractions 模块。

**Comment by Yellow251** [comment]
有点机翻味，但可食用，thanks！

[a comment by an automated account is not shown]

**Reply by liucheng532** [reply]
这一节说实话没怎么看懂

**Comment by Herocly** [comment]
说实话这节有点看不懂

**Comment by Simeon-Chen** [comment]
数据抽象（Data Abstraction）是设计中的一个概念，主要是为了隐藏内部细节，只向外部用户暴露必要的数据信息。它强调重要的「做什么」而不是「如何做」，使用户可以集中注意力于如何使用数据，而不必关心其内部实现的细节。

[a comment by an automated account is not shown]

**Comment by hyazinthekk** [comment]
学到这有种悟道的感觉，这简直是计算机能用0和1模拟整个世界所依靠的精髓所在，堪称人类最伟大的思想之一！

概括来说，这种思路类似于：“如果一个动物看起来像鸭子，走起来像鸭子，叫起来像鸭子，则它就是鸭子。”数据抽象通过“一厢情愿”，绕开了什么是“有理数的本质”的问题，只关注如何实现，最终达到了接近完美的有理数模拟效果。

进一步来说，计算机科学发展至今，通过一层一层抽象，从电路的通断模拟“0和1”，到“2进制数”、到“有理数”，直到“文字”、“图像”、“声音”，曾经被认为不可精确捉摸的东西，都逐渐被成功模拟出来，直到接近“智能”、“情感”……

许多哲学家说人工智能无法具有“真正的”智能，但是什么是智能的本质又难以说清。Dijkstra有一句话，“问机器能不能思考，就像问潜水艇能不能游泳一样”，通过绕开“什么是XXX的本质”的哲学问题，计算机用0和1创造了一个虽然虚拟，但同样“真实”的世界，以至于虚拟和真实、机器传感器与人的感官的的界限，越来越成为被哲学拷问的问题。而只知道自顾自执行自己的规则，看不透“抽象屏障”的上层数据们，就像我们自己一样，谁又能肯定我们的宇宙不是被某个地外文明的超级计算机模拟出来的呢？

有感而发，大家见笑。

**Comment by keyworld32** [comment]
return 错了，应该是return[balabala]

**Comment by wangding** [comment]
```python
>>> from fractions import gcd
>>> def rational(n, d):
        g = gcd(n, d)
        return (n//g, d//g)
```
现在的 Python 版本(>3.9) 应该改成下面
```python
>>> from math import gcd
>>> def rational(n, d):
        g = gcd(n, d)
        return (n//g, d//g)
```

**Comment by lemonacidl** [comment]
个人感觉抽象屏障指的是抽象层级间的分隔。在2.2.3表中，最底部是实现的细节，中间是对数据的基本表示与基本操作。最底部要关注实现细节，而中间的只需要知晓各个函数的作用，抽象的层次更高了些。而在实际编写中，最底部的实现细节可以有很多种，如果后续需要优化（例如性能更好的实现底层方法），在遵守抽象屏障的情况下，对实现细节的修改并不会影响其他层级，便于更新。
        抽象就是从中把握某些性质，在后续的推理中只需要这些细节即可。例如，实现一台电脑。最底层可以看作电路的设计，往上是实现各种原件，原件再组装成各种零件，如CPU、内存之类的。在每个抽象层次中，我或许不知道它的实现细节，但我已知它的功能，我就能对其进行操作。
       如果逾越抽象屏障，例如你在组装电脑的时候薅几个电子元器件，和配件捣鼓在一起，结果成功运行了（虽然像玩笑，但大概这个道理）。能用，但后续的升级维护就会相当困难。

**Comment by mich-ui80** [comment]
这个写的太好了，视频里老师讲的好多地方很难直接用英文理解出来

**Comment by Aticustito** [comment]
集合论秒了（bushi）
数据抽象：数据和操作的组合，约莫相似于群;不抽象的数据，简单的原生数据之类，约莫相等与简单的集合：）
